pyirf/cut_optimization.py

```diff
@@ -1,19 +1,205 @@
 ]
 
 
+def optimize_cuts(
+    signal,
+    background,
+    reco_energy_bins,
+    multiplicity_cuts,
+    gh_cut_efficiencies,
+    theta_cut_efficiencies,
+    fov_offset_min=0 * u.deg,
+    fov_offset_max=1 * u.deg,
+    alpha=1.0,
+    progress=True,
+    **kwargs
+):
+    """
+    Optimize the gamma/hadronnes, theta and multiplicity cut in every energy bin of reconstructed energy
+    for best sensitivity.
+
+    Parameters
+    ----------
+    signal: astropy.table.QTable
+        event list of simulated signal events.
+        Required columns are `theta`, `reco_energy`, 'weight', `gh_score`
+        No directional (theta) or gamma/hadron cut should already be applied.
+    background: astropy.table.QTable
+        event list of simulated background events.
+        Required columns are `reco_source_fov_offset`, `reco_energy`,
+        'weight', `gh_score`.
+        No directional (theta) or gamma/hadron cut should already be applied.
+    reco_energy_bins: astropy.units.Quantity[energy]
+        Bins in reconstructed energy to use for sensitivity computation
+    gh_cut_efficiencies: np.ndarray[float, ndim=1]
+        The cut efficiencies to scan for best sensitivity.
+    theta_cuts: astropy.table.QTable
+        cut definition of the energy dependent theta cut,
+        e.g. as created by ``calculate_percentile_cut``
+    op: comparison function with signature f(a, b) -> bool
+        The comparison function to use for the gamma hadron score.
+        Returning true means an event passes the cut, so is not discarded.
+        E.g. for gammaness-like score, use `operator.ge` (>=) and for a
+        hadroness-like score use `operator.le` (<=).
+    fov_offset_min: astropy.units.Quantity[angle]
+        Minimum distance from the fov center for background events to be taken into account
+    fov_offset_max: astropy.units.Quantity[angle]
+        Maximum distance from the fov center for background events to be taken into account
+    alpha: float
+        Size ratio of off region / on region. Will be used to
+        scale the background rate.
+    progress: bool
+        If True, show a progress bar during cut optimization
+    **kwargs are passed to ``calculate_sensitivity``
+    """
+    gh_cut_efficiencies = np.asanyarray(gh_cut_efficiencies)
+    gh_cut_percentiles = 100 * (1 - gh_cut_efficiencies)
+    fill_value = signal['gh_score'].max()
+
+
+    sensitivities = []
+    cut_indicies = []
+    n_theta_cuts = len(theta_cut_efficiencies)
+    n_gh_cuts = len(gh_cut_efficiencies)
+    n_cuts = len(multiplicity_cuts) * n_theta_cuts * n_gh_cuts
+
+    signal_bin_index, signal_valid = calculate_bin_indices(
+        signal['reco_energy'], reco_energy_bins,
+    )
+    background_bin_index, background_valid = calculate_bin_indices(
+        background['reco_energy'], reco_energy_bins,
+    )
+
+    with tqdm(total=n_cuts, disable=not progress) as bar:
+        for multiplicity_index, multiplicity_cut in enumerate(multiplicity_cuts):
```

Comment:
It might be useful to eventually adapt this to use [scipy.optimize.brute](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.brute.html) internally rather than implement the grid-search manually, as that would make it easier to replace it with more efficient optimizers later

Reply:
Using the manual grid search, we can be a bit cleverer than brute force: make the cuts in the outermost loop possible rather than computing all cuts for all computations.